operator/src/main/java/oracle/kubernetes/operator/http/HttpClient.java

```diff
@@ -249,44 +248,54 @@ public static HttpClient createAuthenticatedClient(final byte[] username, final
    *
-   * @param name The name of the Service that you want the URL for.
-   * @param namespace The Namespace in which the Service you want the URL for is defined.
-   * @return The URL of the Service, or null if it is not found
+   * @param service The name of the Service that you want the URL for.
+   * @return The URL of the Service or null if the URL cannot be found.
    */
-  public static String getServiceURL(String name, String namespace) {
-    if (SERVICE_URL != null) {
-      return SERVICE_URL;
-    }
-    try {
-      CallBuilderFactory factory =
-          ContainerResolver.getInstance().getContainer().getSPI(CallBuilderFactory.class);
-      return getServiceURL(factory.create().readService(name, namespace));
-    } catch (ApiException e) {
-      LOGGER.warning(MessageKeys.EXCEPTION, e);
-    }
-    return null;
+  public static String getServiceURL(V1Service service) {
+    return getServiceURL(service, null, null);
   }
 
   /**
-   * Returns the URL to access the Service; using the Service clusterIP and port.
+   * Returns the URL to access the Service; using the Service clusterIP and port. If the service is
+   * headless, then the pod's IP is returned, if available.
    *
    * @param service The name of the Service that you want the URL for.
+   * @param pod The pod for headless services
+   * @param adminPort administration port name
    * @return The URL of the Service or null if the URL cannot be found.
    */
-  public static String getServiceURL(V1Service service) {
+  public static String getServiceURL(V1Service service, V1Pod pod, String adminPort) {
     if (service != null) {
       V1ServiceSpec spec = service.getSpec();
       if (spec != null) {
-        String portalIP =
-            "None".equalsIgnoreCase(spec.getClusterIP())
-                ? service.getMetadata().getName()
+        String portalIP = spec.getClusterIP();
+        if ("None".equalsIgnoreCase(spec.getClusterIP())) {
+          if (pod != null && pod.getStatus().getPodIP() != null) {
+            portalIP = pod.getStatus().getPodIP();
+          } else {
+            portalIP =
+                service.getMetadata().getName()
                     + "."
                     + service.getMetadata().getNamespace()
-                    + ".svc.cluster.local"
-                : spec.getClusterIP();
-        int port = spec.getPorts().iterator().next().getPort();
+                    + ".pod.cluster.local";
+          }
+        }
+        int port = -1; // uninitialized
+        if (adminPort != null) {
+          for (V1ServicePort sp : spec.getPorts()) {
+            if (adminPort.equals(sp.getName())) {
+              port = sp.getPort();
+              break;
+            }
+          }
+          if (port == -1) {
+            return null;
+          }
+        } else {
+          port = spec.getPorts().iterator().next().getPort();
+        }
         portalIP += ":" + port;
         String serviceURL = HTTP_PROTOCOL + portalIP;
```

Comment:
Just an FYI:  it looks like HTTP_PROTOCOL is still hard coded, I think there's a JIRA out there somewhere to fix this in case it needs to be SSL instead of plain-text.

Reply:
Looking at your other changes, I suspect this needs to be fixed.  The current update detects admin port/channel/default-secure channel should be used, and I suspect all of these can require https.